### PR TITLE
[ATL][ATL_APITEST] Add <atlconv.h> and atlconv testcase

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-16668](https://jira.reactos.org/browse/CORE-16668)

## Proposed changes

- Add `<atlconv.h>` header file.
- Add `atlconv` testcase to `atl_apitest`

## TODO

- [x] Do tests.

## Screenshots

ATL of Visual Studio 2015:

Successful.

Win10:
image
One failure.

### CODE REVIEW

**Comment by katahiromz** [comment]
@learn-more, please have a look of this? https://jira.reactos.org/secure/attachment/66084/atlconv-reference.zip

File: sdk/lib/atl/atlconv.h [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+    LPCSTR m_psz;
+
+    CA2CAEX(_In_z_ LPCSTR psz) : m_psz(psz) { }
+
+    CA2CAEX(_In_z_ LPCSTR psz, _In_ UINT nCodePage) : m_psz(psz)
+    {
+        UNREFERENCED_PARAMETER(nCodePage);
+    }
+
+    ~CA2CAEX() throw() { }
```
Comment (learn-more) [inline sdk/lib/atl/atlconv.h:31]: I'd add a comment both here and in the ctor that this class does not _own_ the string, so there is nothing to free here.

File: sdk/lib/atl/atlconv.h [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+    _Ret_z_ operator LPCSTR() const throw() { return m_psz; }
+
+private:
+    // CA2CAEX is not copyable
+    CA2CAEX(_In_ const CA2CAEX&) throw();
```
Comment (learn-more) [inline sdk/lib/atl/atlconv.h:37]:
```suggestion
    CA2CAEX(_In_ const CA2CAEX&) throw() = delete;
```

File: sdk/lib/atl/atlconv.h [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+
+private:
+    // CW2AEX is not copyable
+    CW2AEX(_In_ const CW2AEX&) throw();
+    CW2AEX& operator=(_In_ const CW2AEX&) throw();
+
+    void Init(_In_z_ LPCWSTR psz, _In_ UINT nConvertCodePage)
+    {
+        int cchMax = WideCharToMultiByte(nConvertCodePage, 0, psz, -1, NULL, 0, NULL, NULL);
+        if (cchMax <= t_nBufferLength)
```
Comment (HBelusca) [inline sdk/lib/atl/atlconv.h:264]: (Note that `_countof(m_szBuffer)` equals `t_nBufferLength`)

File: sdk/lib/atl/atlconv.h [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    CA2AEX(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz);
+        UNREFERENCED_PARAMETER(nCodePage);
+    }
+
+    ~CA2AEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
```
[the remark this answers is not in the document]
Comment (learn-more) [reply sdk/lib/atl/atlconv.h:85]: it does here.

File: sdk/lib/atl/atlconv.h [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+        m_psz[cchMax - 1] = 0;
+    }
+};
+
+typedef CA2AEX<128> CA2A;
+typedef CW2AEX<128> CW2A;
+typedef CA2WEX<128> CA2W;
+typedef CW2WEX<128> CW2W;
+typedef CA2CAEX<128> CA2CA;
+typedef CW2CWEX<128> CW2CW;
```
Comment (learn-more) [inline sdk/lib/atl/atlconv.h:290]: I *think* that if you drop the 128 here, it will use the default from the template.

File: modules/rostests/apitests/atl/atlconv.cpp [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+#ifdef HAVE_APITEST
+    #include <apitest.h>
+#else
+    #include "atltest.h"
+#endif
+#include <atlbase.h>
+#include <atlconv.h>
+
+START_TEST(atlconv)
```
Comment (learn-more) [inline modules/rostests/apitests/atl/atlconv.cpp:16]: If you use the `EX` version, with a template size of something like `5` or at least shorter than your default, does it work as well?

File: sdk/lib/atl/atlconv.h (hunk 3 above)
Comment (katahiromz) [reply sdk/lib/atl/atlconv.h]: Fixed.

File: sdk/lib/atl/atlconv.h [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+            return;
+        }
+
+        m_szBuffer[0] = 0;
+        m_psz = (LPSTR)malloc(cchMax * sizeof(CHAR));
+        if (!m_psz)
+            AtlThrow(E_OUTOFMEMORY);
+
+        WideCharToMultiByte(nConvertCodePage, 0, psz, -1, m_psz, cchMax, NULL, NULL);
+        m_psz[cchMax - 1] = 0;
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/lib/atl/atlconv.h]: Fixed.

File: sdk/lib/atl/atlconv.h (hunk 2 above)
Comment (katahiromz) [reply sdk/lib/atl/atlconv.h]: Fixed.

File: sdk/lib/atl/atlconv.h (hunk 1 above)
Comment (katahiromz) [reply sdk/lib/atl/atlconv.h]: Added comments.

File: sdk/lib/atl/atlconv.h (hunk 5 above)
Comment (katahiromz) [reply sdk/lib/atl/atlconv.h]: Fixed by using `using`.

File: sdk/lib/atl/atlconv.h [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+
+private:
+    // CA2WEX is not copyable
+    CA2WEX(_In_ const CA2WEX&) throw() = delete;
+    CA2WEX& operator=(_In_ const CA2WEX&) throw() = delete;
+
+    void Init(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        int cchMax = MultiByteToWideChar(nCodePage, 0, psz, -1, NULL, 0);
+        if (cchMax <= (int)_countof(m_szBuffer))
```
Comment (HBelusca) [inline sdk/lib/atl/atlconv.h:212]: I wonder what happens if `cchMax` happens to be equal to zero (due to whatever reason: e.g. empty string being converted etc.)...

File: sdk/lib/atl/atlconv.h [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+        {
+            // Allocate a new buffer
+            m_szBuffer[0] = 0;
+            m_psz = (LPWSTR)malloc(cchMax * sizeof(WCHAR));
+            if (!m_psz)
+                AtlThrow(E_OUTOFMEMORY);
+        }
+
+        MultiByteToWideChar(nCodePage, 0, psz, -1, m_psz, cchMax);
+        m_psz[cchMax - 1] = 0;
```
Comment (HBelusca) [inline sdk/lib/atl/atlconv.h:228]: ...especially regarding this part

File: sdk/lib/atl/atlconv.h [hunk 10]
```diff
@@ -0,0 +1,317 @@
...
+    }
+};
+
+using CA2A = CA2AEX<>;
+using CW2A = CW2AEX<>;
+using CA2W = CA2WEX<>;
+using CW2W = CW2WEX<>;
+using CA2CA = CA2CAEX<>;
+using CW2CW = CW2CWEX<>;
```
Comment (HBelusca) [inline sdk/lib/atl/atlconv.h:293]: If `typedef CW2CWEX<> CW2CW;` were used there (for example), this wouldn't compile?

Comment (katahiromz) [reply sdk/lib/atl/atlconv.h:313]: Success. Reverting to `typedef`...

Comment (katahiromz) [reply sdk/lib/atl/atlconv.h]: Fixed.

File: sdk/lib/atl/atlconv.h (hunk 5 above)
Comment (katahiromz) [reply sdk/lib/atl/atlconv.h]: Used `typedef`.

File: sdk/lib/atl/atlconv.h [hunk 11]
```diff
@@ -46,9 +46,9 @@ class CW2CWEX
 public:
     LPCWSTR m_psz;
 
-    CW2CWEX(_In_z_ LPCWSTR psz) : m_psz(psz) { }
+    CW2CWEX(_In_z_ LPCWSTR psz) { }
 
-    CW2CWEX(_In_z_ LPCWSTR psz, _In_ UINT nCodePage) : m_psz(psz)
+    CW2CWEX(_In_z_ LPCWSTR psz, _In_ UINT nCodePage)
```
Comment (HBelusca) [inline sdk/lib/atl/atlconv.h:51]: Why shouldn't `m_psz` be initialized there?

File: sdk/lib/atl/atlconv.h [hunk 12]
```diff
@@ -0,0 +1,341 @@
...
+
+    CA2AEX(_In_z_ LPCSTR psz)
+    {
+        Init(psz);
+    }
+
+    CA2AEX(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz);
+        UNREFERENCED_PARAMETER(nCodePage);
```
Comment (HBelusca) [inline sdk/lib/atl/atlconv.h:81]:
```suggestion
        UNREFERENCED_PARAMETER(nCodePage);
        Init(psz);
```

File: sdk/lib/atl/atlconv.h [hunk 13]
```diff
@@ -0,0 +1,341 @@
...
+
+    CW2WEX(_In_z_ LPCWSTR psz)
+    {
+        Init(psz);
+    }
+
+    CW2WEX(_In_z_ LPCWSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz);
+        UNREFERENCED_PARAMETER(nCodePage);
```
Comment (HBelusca) [inline sdk/lib/atl/atlconv.h:142]:
```suggestion
        UNREFERENCED_PARAMETER(nCodePage);
        Init(psz);
```

File: sdk/lib/atl/atlconv.h (hunk 8 above)
Comment (katahiromz) [reply sdk/lib/atl/atlconv.h:230]: Added the tests of zero-length.

File: modules/rostests/apitests/atl/atlconv.cpp (hunk 6 above)
Comment (katahiromz) [reply modules/rostests/apitests/atl/atlconv.cpp:16]: Added.

File: sdk/lib/atl/atlconv.h [hunk 14]
```diff
@@ -322,7 +322,7 @@ class CW2AEX
             }
 
             // A complex UTF-8 string might come here
-            cchMax = 3 * cch + 1;
+            cchMax = 4 * cch + 1;
```
Comment (HBelusca) [inline sdk/lib/atl/atlconv.h:325]: Doesn't WideCharToMultiByte return the expected number of WCHAR units for the string?
> If successful, returns the number of bytes written to the buffer pointed to by lpMultiByteStr. _**If the function succeeds and cbMultiByte is 0, the return value is the required size, in bytes, for the buffer indicated by lpMultiByteStr.**_

Comment (katahiromz) [reply sdk/lib/atl/atlconv.h]: Fixed by using `WideCharToMultiByte`.

File: sdk/lib/atl/atlconv.h [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+    void Init(_In_z_ LPCWSTR psz, _In_ UINT nConvertCodePage)
+    {
+        if (!psz)
+        {
+            m_psz = NULL;
+            m_szBuffer[0] = 0;
+            return;
+        }
+
+        // Calculation of WideCharToMultiByte is slow. Use lstrlenW instead.
```
Comment (learn-more) [inline sdk/lib/atl/atlconv.h:294]: Was this a problem while testing?
I'd prefer if there were no shortcuts like this unless it is actually needed, because increased complexity leads to more bugs.

Comment (katahiromz) [reply sdk/lib/atl/atlconv.h:294]: > Was this a problem while testing?

Yep. There's a corner case against this implementation. It looks like an optimization.

Comment (katahiromz) [reply sdk/lib/atl/atlconv.h:294]: The failure like this:
```txt
atlconv.cpp:125: Test failed: w2ax.m_psz == w2ax.m_szBuffer
```
happens.

Comment (katahiromz) [reply sdk/lib/atl/atlconv.h:303]: Some #ifdef's added. Please compare these codes.

File: modules/rostests/apitests/atl/atlconv.cpp [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+    {
+        CA2AEX<5> a2ax("123");
+        ok_str((LPSTR)a2ax, "123");
+        ok_str(a2ax.m_psz, "123");
+        ok_ptr(a2ax.m_psz, a2ax.m_szBuffer);
+
+        CW2AEX<5> w2ax(L"123");
+        ok_str((LPSTR)w2ax, "123");
+        ok_str(w2ax.m_psz, "123");
+        ok(w2ax.m_psz != w2ax.m_szBuffer, "w2ax.m_psz == w2ax.m_szBuffer\n"); // corner case
```
Comment (katahiromz) [inline modules/rostests/apitests/atl/atlconv.cpp:125]: This is the corner case I found.

File: sdk/lib/atl/atlconv.h [hunk 17]
```diff
@@ -227,8 +227,13 @@ class CA2WEX
             return;
         }
 
+#if 1
+        // This is slow
```
[the remark this answers is not in the document]
Comment (katahiromz) [reply sdk/lib/atl/atlconv.h:231]:
```c
#include <windows.h>
#include <stdio.h>

int main(void)
{
    const char *psz = "MultiByteToWideChar(nCodePage, 0, , -1, NULL, 0);";

    DWORD dwTick0 = GetTickCount();
    for (DWORD i = 0; i < 10000000; ++i)
        MultiByteToWideChar(CP_ACP, 0, psz, -1, NULL, 0);
    DWORD dwTick1 = GetTickCount();
    INT cch2 = 0;
    for (DWORD i = 0; i < 10000000; ++i)
    {
        INT cch = lstrlenA(psz);
        cch2 += cch;
    }
    DWORD dwTick2 = GetTickCount();
    printf("%ld\n", dwTick1 - dwTick0);
    printf("%ld\n", dwTick2 - dwTick1);
    printf("%d\n", cch2);
    return 0;
}
```

image
It's x3 fast.

File: sdk/lib/atl/atlconv.h [hunk 18]
```diff
@@ -0,0 +1,348 @@
...
+    void Init(_In_z_ LPCWSTR psz, _In_ UINT nConvertCodePage)
+    {
+        if (!psz)
+        {
+            m_psz = NULL;
+            m_szBuffer[0] = 0;
+            return;
+        }
+
+        // NOTE: This has a failure.
```
Comment (HBelusca) [inline sdk/lib/atl/atlconv.h:297]: This comment doesn't explain anything.

Comment (katahiromz) [reply sdk/lib/atl/atlconv.h:297]:
```c
// NOTE: This code is a cause of a failure on `atl_apitest atlconv` testcase. FIXME.
```
